src/plots/cartesian/index.js

```diff
@@ -145,4 +147,211 @@ exports.clean = function(newFullData, newFullLayout, oldFullData, oldFullLayout)
+        joinLayer(parent, 'g', 'scatterlayer');
+    }
+
+    if(!plotinfo.mainplot) {
+        plotinfo.bg = joinLayer(plotgroup, 'rect', 'bg');
+        plotinfo.bg.style('stroke-width', 0);
+
+        var backLayer = joinLayer(plotgroup, 'g', 'layer-subplot');
+        plotinfo.shapelayer = joinLayer(backLayer, 'g', 'shapelayer');
+        plotinfo.imagelayer = joinLayer(backLayer, 'g', 'imagelayer');
+
+        plotinfo.gridlayer = joinLayer(plotgroup, 'g', 'gridlayer');
+        plotinfo.overgrid = joinLayer(plotgroup, 'g', 'overgrid');
+
+        plotinfo.zerolinelayer = joinLayer(plotgroup, 'g', 'zerolinelayer');
+        plotinfo.overzero = joinLayer(plotgroup, 'g', 'overzero');
+
+        plotinfo.plot = joinLayer(plotgroup, 'g', 'plot');
+        plotinfo.overplot = joinLayer(plotgroup, 'g', 'overplot');
+
+        plotinfo.xlines = joinLayer(plotgroup, 'path', 'xlines');
+        plotinfo.ylines = joinLayer(plotgroup, 'path', 'ylines');
+        plotinfo.overlines = joinLayer(plotgroup, 'g', 'overlines');
+
+        plotinfo.xaxislayer = joinLayer(plotgroup, 'g', 'xaxislayer');
+        plotinfo.yaxislayer = joinLayer(plotgroup, 'g', 'yaxislayer');
+        plotinfo.overaxes = joinLayer(plotgroup, 'g', 'overaxes');
+    }
+    else {
+
+        // now make the components of overlaid subplots
+        // overlays don't have backgrounds, and append all
+        // their other components to the corresponding
+        // extra groups of their main plots.
+
+        var mainplot = fullLayout._plots[plotinfo.mainplot];
+        mainplot.overlays.push(plotinfo);
+
+        plotinfo.gridlayer = joinLayer(mainplot.overgrid, 'g', subplot);
+        plotinfo.zerolinelayer = joinLayer(mainplot.overzero, 'g', subplot);
+
+        plotinfo.plot = joinLayer(mainplot.overplot, 'g', subplot);
+        plotinfo.xlines = joinLayer(mainplot.overlines, 'path', subplot);
+        plotinfo.ylines = joinLayer(mainplot.overlines, 'path', subplot);
+        plotinfo.xaxislayer = joinLayer(mainplot.overaxes, 'g', subplot);
+        plotinfo.yaxislayer = joinLayer(mainplot.overaxes, 'g', subplot);
+    }
+
+    // common attributes for all subplots, overlays or not
+    plotinfo.plot.call(plotLayers);
+
+    plotinfo.xlines
+        .style('fill', 'none')
+        .classed('crisp', true);
+
+    plotinfo.ylines
+        .style('fill', 'none')
+        .classed('crisp', true);
+}
+
+function purgeSubplotLayers(layers, fullLayout) {
+    if(!layers) return;
+
+    layers.each(function(subplot) {
+        var plotgroup = d3.select(this),
+            clipId = 'clip' + fullLayout._uid + subplot + 'plot';
+
+        plotgroup.remove();
+        fullLayout._draggers.selectAll('g.' + subplot).remove();
+        fullLayout._defs.select('#' + clipId).remove();
+
+        // do not remove individual axis <clipPath>s here
+        // as other subplots may need them
+    });
+}
+
+
+function joinLayer(parent, nodeType, className) {
```

Comment:
Yes, def. We use this pattern all over the place, and it's invalid in d3 v4 without some minor modifications.